include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -132,11 +140,43 @@ struct _HasDefaultName<_ExternalName<_InternalName>>
     static constexpr bool value = _HasDefaultName<_InternalName>::value;
 };
 
+template <typename... _Name>
+struct __optional_kernel_name;
+
+template <typename _CustomName>
+using __kernel_name_provider =
+#if __SYCL_UNNAMED_LAMBDA__
+    typename ::std::conditional<_HasDefaultName<_CustomName>::value, __optional_kernel_name<>,
+                                __optional_kernel_name<_CustomName>>::type;
+#else
+    __optional_kernel_name<_CustomName>;
+#endif
+
+template <char...>
+struct __composite_kernel_name
+{
+};
+
+template <typename T>
+class __kernel_name_composer
+{
+    static constexpr auto __name = __builtin_sycl_unique_stable_name(T);
+    static constexpr std::size_t __name_size = __builtin_strlen(__name);
+
+    template <::std::size_t... Is>
+    static __composite_kernel_name<__name[Is]...>
+    __compose_kernel_name(::std::index_sequence<Is...>);
```

Comment:
I ran some benchmark using Intel® VTune™ and got the following kernel name:

before patch:
```c++
__parallel_scan_local_kernel<oneapi::dpl::execution::__dpl::DefaultKernelName, oneapi::dpl::__ranges::all_view<float, (cl::sycl::access::mode)1024, (cl::sycl::access::target)2014, (cl::sycl::access::placeholder)1>, oneapi::dpl::__ranges::all_view<float, (cl::sycl::access::mode)1025, (cl::sycl::access::target)2014, (cl::sycl::access::placeholder)1>, std::plus<float>, float, oneapi::dpl::unseq_backend::__scan<std::integral_constant<bool, (bool)0>, oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, std::plus<float>, oneapi::dpl::unseq_backend::walk_n<oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, oneapi::dpl::__internal::__no_op>, oneapi::dpl::unseq_backend::__scan_assigner, oneapi::dpl::unseq_backend::__scan_assigner, oneapi::dpl::unseq_backend::walk_n<oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, oneapi::dpl::__internal::__no_op>, oneapi::dpl::unseq_backend::__scan_init<float>>, oneapi::dpl::unseq_backend::__scan<std::integral_constant<bool, (bool)1>, oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, std::plus<float>, oneapi::dpl::unseq_backend::walk_n<oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, oneapi::dpl::__internal::__no_op>, oneapi::dpl::unseq_backend::__scan_no_assign, oneapi::dpl::unseq_backend::__scan_assigner, oneapi::dpl::unseq_backend::walk_n<oneapi::dpl::execution::__dpl::device_policy<oneapi::dpl::execution::__dpl::DefaultKernelName> const&, oneapi::dpl::__internal::__no_op>, oneapi::dpl::unseq_backend::__scan_no_init<float>>, oneapi::dpl::unseq_backend::__global_scan_functor<std::integral_constant<bool, (bool)0>, std::plus<float>>>
```

after patch:
```c++
__composite_kernel_name<(char)95, (char)90, (char)84, (char)83, (char)78, (char)54, (char)111, (char)110, (char)101, (char)97, (char)112, (char)105, (char)51, (char)100, (char)112, (char)108, (char)50, (char)48, (char)95, (char)95, (char)112, (char)97, (char)114, (char)95, (char)98, (char)97, (char)99, (char)107, (char)101, (char)110, (char)100, (char)95, (char)104, (char)101, (char)116, (char)101, (char)114, (char)111, (char)50, (char)56, (char)95, (char)95, (char)112, (char)97, (char)114, (char)97, (char)108, (char)108, (char)101, (char)108, (char)95, (char)115, (char)99, (char)97, (char)110, (char)95, (char)108, (char)111, (char)99, (char)97, (char)108, (char)95, (char)107, (char)101, (char)114, (char)110, (char)101, (char)108, (char)73, (char)74, (char)78, (char)83, (char)48, (char)95, (char)57, (char)101, (char)120, (char)101, (char)99, (char)117, (char)116, (char)105, (char)111, (char)110, (char)53, (char)95, (char)95, (char)100, (char)112, (char)108, (char)49, (char)55, (char)68, (char)101, (char)102, (char)97, (char)117, (char)108, (char)116, (char)75, (char)101, (char)114, (char)110, (char)101, (char)108, (char)78, (char)97, (char)109, (char)101, (char)69, (char)78, (char)83, (char)48, (char)95, (char)56, (char)95, (char)95, (char)114, (char)97, (char)110, (char)103, (char)101, (char)115, (char)56, (char)97, (char)108, (char)108, (char)95, (char)118, (char)105, (char)101, (char)119, (char)73, (char)102, (char)76, (char)78, (char)50, (char)99, (char)108, (char)52, (char)115, (char)121, (char)99, (char)108, (char)54, (char)97, (char)99, (char)99, (char)101, (char)115, (char)115, (char)52, (char)109, (char)111, (char)100, (char)101, (char)69, (char)49, (char)48, (char)50, (char)52, (char)69, (char)76, (char)78, (char)83, (char)65, (char)95, (char)54, (char)116, (char)97, (char)114, (char)103, (char)101, (char)116, (char)69, (char)50, (char)48, (char)49, (char)52, (char)69, (char)76, (char)78, (char)83, (char)65, (char)95, (char)49, (char)49, (char)112, (char)108, (char)97, (char)99, (char)101, (char)104, (char)111, (char)108, (char)100, (char)101, (char)114, (char)69, (char)49, (char)69, (char)69, (char)69, (char)78, (char)83, (char)55, (char)95, (char)73, (char)102, (char)76, (char)83, (char)66, (char)95, (char)49, (char)48, (char)50, (char)53, (char)69, (char)76, (char)83, (char)67, (char)95, (char)50, (char)48, (char)49, (char)52, (char)69, (char)76, (char)83, (char)68, (char)95, (char)49, (char)69, (char)69, (char)69, (char)102, (char)78, (char)83, (char)48, (char)95, (char)49, (char)51, (char)117, (char)110, (char)115, (char)101, (char)113, (char)95, (char)98, (char)97, (char)99, (char)107, (char)101, (char)110, (char)100, (char)54, (char)95, (char)95, (char)115, (char)99, (char)97, (char)110, (char)73, (char)83, (char)116, (char)49, (char)55, (char)105, (char)110, (char)116, (char)101, (char)103, (char)114, (char)97, (char)108, (char)95, (char)99, (char)111, (char)110, (char)115, (char)116, (char)97, (char)110, (char)116, (char)73, (char)98, (char)76, (char)98, (char)48, (char)69, (char)69, (char)82, (char)75, (char)78, (char)83, (char)52, (char)95, (char)49, (char)51, (char)100, (char)101, (char)118, (char)105, (char)99, (char)101, (char)95, (char)112, (char)111, (char)108, (char)105, (char)99, (char)121, (char)73, (char)83, (char)53, (char)95, (char)69, (char)69, (char)83, (char)116, (char)52, (char)112, (char)108, (char)117, (char)115, (char)73, (char)102, (char)69, (char)78, (char)83, (char)71, (char)95, (char)54, (char)119, (char)97, (char)108, (char)107, (char)95, (char)110, (char)73, (char)83, (char)78, (char)95, (char)78, (char)83, (char)48, (char)95, (char)49, (char)48, (char)95, (char)95, (char)105, (char)110, (char)116, (char)101, (char)114, (char)110, (char)97, (char)108, (char)55, (char)95, (char)95, (char)110, (char)111, (char)95, (char)111, (char)112, (char)69, (char)69, (char)69, (char)78, (char)83, (char)71, (char)95, (char)49, (char)53, (char)95, (char)95, (char)115, (char)99, (char)97, (char)110, (char)95, (char)97, (char)115, (char)115, (char)105, (char)103, (char)110, (char)101, (char)114, (char)69, (char)83, (char)85, (char)95, (char)83, (char)84, (char)95, (char)78, (char)83, (char)71, (char)95, (char)49, (char)49, (char)95, (char)95, (char)115, (char)99, (char)97, (char)110, (char)95, (char)105, (char)110, (char)105, (char)116, (char)73, (char)102, (char)69, (char)69, (char)69, (char)69, (char)69, (char)69, (char)69>
```
It was not readable enough before the patch, but now...

Reply:
yes and I know that. I guess we have no choice. Anyway, even if it was readable user cannot match it on his/her code because they are still implementation details. If user wants something meaningful and guaranteed not to be change between releases the explicit kernel name shall be set